src/core/document.js

```diff
@@ -308,29 +309,40 @@ var Page = (function PageClosure() {
     },
 
     getAnnotationsData: function Page_getAnnotationsData(intent) {
-      var annotations = this.annotations;
-      var annotationsData = [];
-      for (var i = 0, n = annotations.length; i < n; ++i) {
-        if (!intent || isAnnotationRenderable(annotations[i], intent)) {
-          annotationsData.push(annotations[i].data);
+      var annotationsPromise = this.annotations ?
+        this.annotations : this.getAnnotations();
+      return annotationsPromise.then(function (annotations) {
+        var annotationsData = [];
+        for (var i = 0, n = annotations.length; i < n; ++i) {
+          if (!intent || isAnnotationRenderable(annotations[i], intent)) {
+            annotationsData.push(annotations[i].data);
+          }
         }
-      }
-      return annotationsData;
+        return annotationsData;
+      });
     },
 
-    get annotations() {
-      var annotations = [];
+    getAnnotations: function Page_getAnnotations() {
       var annotationRefs = this.getInheritedPageProp('Annots') || [];
+      var annotationsPromises = [];
       for (var i = 0, n = annotationRefs.length; i < n; ++i) {
         var annotationRef = annotationRefs[i];
-        var annotation = AnnotationFactory.create(this.xref, annotationRef,
-                                                  this.pdfManager,
-                                                  this.idFactory);
-        if (annotation) {
-          annotations.push(annotation);
-        }
+        var annotationPromise = AnnotationFactory.create(
+          this.xref,
+          annotationRef,
+          this.pdfManager,
+          this.idFactory
+        );
+
+        annotationsPromises.push(annotationPromise);
       }
-      return shadow(this, 'annotations', annotations);
+
+      return shadow(this, 'annotations',
```

Comment:
No need for the shadowing anymore now that this is not a getter anymore. You can change it to `this._annotations = Promise.all(...). then(...);` and then `return this._annotations;`.